### PR TITLE
CO: Proposal for PrestaShop Calendars - STEP 1/3 - Calendars Base Section

### PR DESCRIPTION
**Pull request by AliShareei**
Proposal for PrestaShop Calendars - STEP 1/3 - Calendars Base Section

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR is a Proposal method for the Prestashop calendars. this step is base step. step 2 is about DatePickers and Step 3 is about Convert Date when submit forms (for save in DB). I'm interested in talking about this.
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-4642
| How to test?  | Please enter the SQL code inside the file 1.7.4.0.sql into the database, then log in to BackOffice and select your calendar from the language settings or employee settings.

#### Important notice
* This method is flexible and usable for developing and adding calendars and this proposal method, according to a few months ago, has been done on this page, which I think is close to what we want. https://trello.com/c/TnGYz4pZ/59-add-jalali-calendar
* This method does not require an extension or library separation.
* this is first Step from 3 Step for Calendars
* Step 2 is about DatePickers
* Step 3 is about Convert Date when submit forms with date pickers.
* This is a suggested approach to start, and we can modify and use it according by suggestions and improvements.
* The database code was placed inside the file 1.7.4.0.sql, I did not know where and which version I should put!
* Let's get started.

Thank you.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8971)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Interesting, we need help from someone part of Product management here :grinning:

File: classes/calendar/Calendar.php [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    public static function getCalendarsList($showDefaultOption = false)
+    {
+        $translator = Context::getContext()->getTranslator();
+        $list = array();
+		
+		if ($showDefaultOption) {
+			$list[] = array(
+				'id' => 0,
+				'name' => $translator->trans('-- Based on language settings --', array(), 'Admin.Global'),
```
Comment (LouiseBonnard) [inline classes/calendar/Calendar.php:113]: I'm not sure of the action to do here, is it a drop-down list? If yes, perhaps it is more relevant to write _-- Please choose your language --_ ou _-- Choose --_ (Admin/Actions). What do you think?

File: classes/calendar/Calendar.php [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			);
+		}
+    
+        foreach (static::$calendars as $id => $name) {
+            $list[] = array(
+                'id' => $id,
+                'name' => $translator->trans(sprintf('%s calendar', $name), array(), 'Admin.Global')
```
Comment (LouiseBonnard) [inline classes/calendar/Calendar.php:120]: Same here, do we agree on the fact that there are two data here: country calendar + its name? Why not only the name of the calendar?

File: classes/calendar/Calendar.php (hunk 1 above)
Comment (AliShareei) [reply classes/calendar/Calendar.php:113]: prestashop-calendars

Hello @LouiseBonnard ,

yes it's drop-down list, in 2 place. first in language settings and second in employee settings.
as you see in image in language setting just we have calendars list in drop-down but in employee settings we have extra option "-- Based on language settings --" this option if choose buy user is meaning Display the calendar according to the language settings for the user.

My English language is not very good, So we can choose another text. But it's better to give the user the idea that if the user does not select a specific calendar (choose this option), Displays the calendar according to the settings specified in the user's language.

so we can use like this 

- Calendar set in language -
- Calendar in language -
- Default Language Calendar - 
- Default by language -
- Language Default -
- Default -

As I said, I do not know which words are better. what is your opinion?

Thanks.

File: classes/calendar/Calendar.php (hunk 2 above)
Comment (AliShareei) [reply classes/calendar/Calendar.php:120]: I agree with you, only the calendar name is good. for example "Gregorian" and "Jalali".

**Comment by AliShareei** [comment]
dear @mickaelandrieu , @LouiseBonnard , @colinegin 

I also say some other points to help you in the review.

1. As you can see in codes, we have changed the Customer class as the same change in Employee class. I was not sure about this item, but we can have the same settings as the Employee settings in the Front Office section (in user account like this image)

prestashop-customer-calendar

2. I tried to implement this method in the simplest possible way. 
But if we have a table for Calendars, we might be able to add more flexibility to the project, for example, with modules, we'll be able to add the calendar to PrestaShop.

| id_calendar   | name           |
| ------------- |:--------------:|
| 1             | Gregorian      |
| 2             | Jalali         |
| 3             | CustomByModule |

Regards,

File: classes/calendar/Calendar.php (hunk 1 above)
Comment (LouiseBonnard) [reply classes/calendar/Calendar.php:113]: Okay then, thanks for this explanation: _-- Based on language setting --_ (Admin.Global). @colinegin, what do you think?

**Comment by colinegin** [comment]
Hi @AliShareei, thanks for your PR !

Just a few questions as I'm not exactly sure what you had decided with Alexandra at the time you discussed about RTL : 
- Should the calendar depend on the country or on the language ? It seems that Alexandra talked about country dependancy. 
- Why don't you add the calendar choice in the localization page ?
- Why is it necessary to add calendar choice in the employee profile ?
- On Front-office, why is it necessary to add calendar choice in the customer account ? We could make it simpler by automatically changing the calendar when the customer chooses another language. 

What do you think ?

Have a nice day :)

**Comment by AliShareei** [comment]
Hello dear @colinegin , you're welcome.

when we talk with Alexandra she make 2 mockups in trello.
https://trello.com/c/TnGYz4pZ/59-add-jalali-calendar

rtl_jalali_mockups

as you see in image first is a configuration in localization Page and second is a dropdown list in employee settings.

about your questions :

- with alexandra we don't talk about country. we just talk about language and first mockup. because i told to him when we have a multi language shop we have a setting for each language.

I do not have bias about language or country. but as you see in PrestaShop language we have full DateFormat and lite DateFormat in language setting so this is related with Calendar. 

for example we have multi language shop. Users can also choose their language in stores and customer and employee can select language in profile.
also every language is related with country for example persian is fa-IR, united stated has en-US.

so i think language setting is better for Calendar setting. But as I said, I do not insist on this issue and this proposed method can be in the country. But my offer is the language setting.

- When the store opens according to the language (or country) of the user, the calendar is set to the same basis, and there is no need for an extra option on other pages like  localization page.

- about employee setting, as you see in second mockups, Alexandra make this option for Employee. so i think this option is good. Because it gives more flexibility to the store manager and Employees. For example, we have a global store that accepts orders from different countries. We may have employees at the store who are responsible for accepting orders from a particular area, so the language of the employee may be the same as the language of the other employees, but will require a different calendar.
That's why I think this option can be good for specific use

- about Front-office, Note that when the language is automatically detected by the browser for the user, the calendar will also be displayed in the same language for the user. So the calendar is also automatically set for it.

We may have a store with a language but customers from around the world maybe they want to use the store on their own calendar. or example, a traveler or a migrant in another country, he may want to choose his own language, but he will follow the calendar of the country where he is located.

These uses are possible in certain circumstances, and so I did not know if such an option would be needed in the front office section,That's why I did not fully implement this section for Front-office, and I just suggested it.  

- In general, I think this method fits well with all the needs and can easily solve Prestashop calendar problems.
These are my comments, but to change it, we can improve it according to what you think.

Thank you and have a good day too.

**Comment by AliShareei** [comment]
dear @colinegin ,

about country or language, 
 
 I add that each country has its own official calendar, and it is reasonable that this setting is in the country table, but I do not know how the customers can choose their country to access their official calendar.

There are selectable settings in the Front-office, one for the language and one for the currency unit.

If there is a way to change the country by the customers in Front-Office, then it might be possible to think about it to change Calendar by country.

But when customers in the front-office can not choose their country, it does not seem appropriate to change the calendar according to the country.

or for example in employee setting, they can not choose country too.

What do you think ?

Thanks.

**Comment by colinegin** [comment]
Hello @AliShareei, 

Thanks for this detailed explanation !

We are still discussing with the rest of the team about what to choose between country and language. Thank you for giving us examples and pros/cons for both options, this is of great help. I will come back to you with a decision as soon as possible.

Regarding calendar choice on front office, I'm quite sure there is no calendar displayed in the core software. This is only by installing specific modules that a calendar can be added on the front office (on product pages for instance). So I don't think this is necessary to add a calendar dropdown in the customer account, as it can be confusing for the customer.

What do you think ?

Again thank you and have a nice evening.

**Comment by AliShareei** [comment]
Hello dear @colinegin ,

about Front-Office, customer account pages order list (order-history) and order Detail (top page and bottom for customer messages) used Date and Calendar, so i think in Front-Office also we have Calendar even without Modules.

In my opinion, the proposed approach would be better to consider FrontOffice, which means that the developed modules also have a specific structure and know how to use the Prestashop core calendar for their own purposes and not to worry about it.

As we know, a single calendar is currently being displayed by selecting different languages and this should also be considered.

Thanks for your review and follow up.
I'm waiting for your final decision to create this PR  and other steps on the basis of your final decisions.

also have a nice day too.

**Comment by MatShir** [comment]
Hello @AliShareei sorry for the waiting, hope you still want to continue this project which looks very promising. 
Just further questions, I really think as well that linking the calendar to the language is the best option. But my only concern is for the front office in the client information, this kind of setting is not common for the customers (could give douts and incomprehension) and also merchants want to minimize clicks number and interactions. 
What do you think? 

Sorry again, we really consider your project but we went through a lot lately (released, new organization and projects...).

**Comment by MatShir** [comment]
#11571 is related need around the Jalali calendar

**Comment by AliShareei** [comment]
Hello @MatShir and thank you for reviewing.
We've been waiting for your feedback for months to keep up.

According to your reviewing and your colleagues.
We can cancel changes on front office (customers). And just add the calendar in the language sections and employee settings.

If the method and structure are appropriate, we can continue and finish it. Is there a better way to communicate with you? Or your technical team? So small technical questions will answer sooner?

**Comment by MatShir** [comment]
Hi @AliShareei, 
Great.  For frequent communication, we could use Gitter on a private channel. I will do the link with the technical team if needed. If you want we can make a call to start properly and ask all your needs before you start again. 
If it is good for you?

**Comment by AliShareei** [comment]
Hi @MatShir ,
Very good!
Because a lot of time has passed, I will first have a review over the coming days, and I'll put it together for you and your colleagues to work on.
I will be contacting you again during the coming days(next week), and if I have any questions or technical ambiguity, I will be speaking at that time.
have a good day.

**Comment by MatShir** [comment]
Hi @AliShareei,
Here the room https://gitter.im/PrestaShop/JalaliCalendar?utm_source=share-link&utm_medium=link&utm_campaign=share-link on the gitter platform.

**Comment by AliShareei** [comment]
Hi @MatShir 
Fortunately, I'm tracking and doing this section and will soon be notified of the next steps.

Thanks for the link. I logged in to PrestaShop's Gitter account. but, the JalaliCalendar room on Gitter had a 404 error. I do not know if this page is missing or I do not have access.

**Comment by PierreRambaud** [comment]
ping @PrestaShop/product-team

**Comment by matks** [comment]
> ping @PrestaShop/product-team

Following the above, I add `Waiting for PM` label as it seems an input from Product team is needed.

**Comment by PierreRambaud** [comment]
Hi @AliShareei ,

Since we had no news from the @PrestaShop/product-team  for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**